src/field_theory/galois_field.lean

```diff
@@ -0,0 +1,214 @@
+splitting_field (X^(p^n) - X : polynomial (zmod p))
+
+instance : inhabited (@galois_field 2 (fact.mk nat.prime_two) 1) :=
+⟨37⟩
+
+namespace galois_field
+variables (p : ℕ) [fact p.prime] (n : ℕ)
+
+instance : algebra (zmod p) (galois_field p n) :=
+splitting_field.algebra _
+
+instance : is_splitting_field (zmod p) (galois_field p n) (X^(p^n) - X) :=
+polynomial.is_splitting_field.splitting_field _
+
+instance : char_p (galois_field p n) p :=
+(algebra.char_p_iff (zmod p) (galois_field p n) p).mp (by apply_instance)
+
+-- should be able to apply_instance from finite_dimensional.fintype on finite.lean
+instance : fintype (galois_field p n) :=
+finite_dimensional.fintype_of_fintype (zmod p) (galois_field p n)
+
+local notation [parsing_only] `g_poly` := (X^(p^n) - X : polynomial (zmod p))
+
+section
+
+variables {K : Type*} [field K] {p n} (hn : n ≠ 0)
+include hn
+
+lemma gpoly_nat_degree_eq : (X ^ p ^ n - X : polynomial K).nat_degree = p ^ n :=
+begin
+  have hp : 1 < p := (fact.out (nat.prime p)).one_lt,
+  have h1 : (X : polynomial K).degree < (X ^ p ^ n : polynomial K).degree,
+  { rw [degree_X_pow, degree_X],
+    exact_mod_cast nat.one_lt_pow _ _ (nat.pos_of_ne_zero hn) (nat.prime.one_lt' p).1 },
+  rw [nat_degree_eq_of_degree_eq (degree_sub_eq_left_of_degree_lt h1), nat_degree_X_pow],
+end
+
+lemma gpoly_ne_zero : (X ^ p ^ n - X : polynomial K) ≠ 0 :=
+ne_zero_of_nat_degree_gt $
+calc 1 < _ : nat.one_lt_pow _ _ (nat.pos_of_ne_zero hn) (nat.prime.one_lt' p).1
+... = _ : (gpoly_nat_degree_eq hn).symm
+
+end
+
+lemma finrank {n} (h : n ≠ 0) : finite_dimensional.finrank (zmod p) (galois_field p n) = n :=
+begin
+  have aux : g_poly ≠ 0 := gpoly_ne_zero h,
+  have key : fintype.card ((g_poly).root_set (galois_field p n)) = (g_poly).nat_degree :=
+    card_root_set_eq_nat_degree (galois_poly_separable p _ (dvd_pow (dvd_refl p) h))
+    (splitting_field.splits g_poly),
+  have nat_degree_eq : (g_poly).nat_degree = p ^ n := gpoly_nat_degree_eq h,
+  rw nat_degree_eq at key,
+  suffices : (g_poly).root_set (galois_field p n) = set.univ,
+  { simp_rw [this, ←fintype.of_equiv_card (equiv.set.univ _)] at key,
+    rw [@card_eq_pow_findim (zmod p), zmod.card] at key,
+    exact nat.pow_right_injective ((nat.prime.one_lt' p).out) key },
+  rw set.eq_univ_iff_forall,
+  suffices : ∀ x (hx : x ∈ (⊤ : subalgebra (zmod p) (galois_field p n))),
+    x ∈ (X ^ p ^ n - X : polynomial (zmod p)).root_set (galois_field p n),
+  { simpa, },
+  rw ← splitting_field.adjoin_root_set,
+  simp_rw algebra.mem_adjoin_iff,
+  intros x hx,
+  apply subring.closure_induction hx; clear_dependent x; simp_rw mem_root_set aux,
+  { rintros x (⟨r, rfl⟩ | hx),
+    { simp only [aeval_X_pow, aeval_X, alg_hom.map_sub],
+      rw [← ring_hom.map_pow, zmod.pow_card_pow, sub_self], },
+    { dsimp only [galois_field] at hx,
+      rwa mem_root_set aux at hx, }, },
+  { simp only [aeval_X_pow, zero_pow_eq_zero, aeval_X, sub_zero, alg_hom.map_sub],
```

Comment:
i was kinda sad to find that `polynomial.coeff_zero_eq_aeval_zero` doesn't work here; is it generalizable or am i missing some obvious technicality

Reply:
I added a generalized version which seems useful